src/bootstrap/src/core/config/config.rs

```diff
@@ -1978,13 +1990,49 @@ impl Config {
             config.ferrocene_aws_profile = f.aws_profile;
             config.ferrocene_oxidos_src = f.oxidos_src;
             config.ferrocene_tarball_signing_kms_key_arn = f.tarball_signing_kms_key_arn;
-            config.ferrocene_document_signatures_s3_bucket = f
-                .document_signatures_s3_bucket
-                .unwrap_or_else(|| "ferrocene-document-signatures".into());
-            config.ferrocene_ignore_document_signatures =
-                f.ignore_document_signatures.unwrap_or(false);
             config.ferrocene_technical_report_url = f.technical_report_url;
 
+            config.ferrocene_document_signatures = match (
+                f.document_signatures.as_deref(),
+                f.document_signatures_s3_bucket,
+                f.document_signatures_tarball,
+            ) {
+                (None | Some("s3"), Some(bucket), None) => {
+                    FerroceneDocumentSignatures::S3 { bucket }
+                }
+                (None | Some("s3"), None, None) => FerroceneDocumentSignatures::S3 {
+                    bucket: "ferrocene-document-signatures".into(),
+                },
+                (None | Some("s3"), _, _) => panic!(
+                    "ferrocene.document-signatures=\"s3\" only supports \
+                     ferrocene.document-signatures-s3-bucket as the additional \
+                     setting related to document signatures"
+                ),
```

Comment:
This is the case when there is a tarball path given, but no s3 bucket, right? It took me a few seconds to think that through. Maybe following is slightly more clear?

```suggestion
                (None | Some("s3"), None, Some(_)) => panic!(
                    "ferrocene.document-signatures=\"s3\" only supports \
                     ferrocene.document-signatures-s3-bucket as the additional \
                     setting related to document signatures"
                ),
```

Reply:
This is the case where you configured the `s3` mode, but you also used configuration keys the `s3` directive doesn't know how to handle. It's written this way to require minimal changes when a new mode will be added in the future.

Right now we have something like this (pseudocode):

```rust
match (mode, s3_bucket, docs_tarball) {
    ("s3",           None,         None         ) => ok(),
    ("s3",           Some(bucket), None         ) => ok(),
    ("s3",           _,          , _            ) => err(),

    ("docs-tarball", None,       , Some(tarball)) => ok(),
    ("docs-tarball", None,       , _            ) => err(),

    ("disabled",     None,       , None         ) => ok(),
    ("disabled",     _,          , _            ) => err(),

    (_,              _,          , _            ) => err(),
}
```

The purpose of the `("mode", _,  _)` lines is to handle all other possibilities that the mode doesn't support.

We *could* right now do (for example) `("s3", None | Some(_), Some(_))`, where the first `None | Some(_)` handles both values for the supported option and `Some(_)` handles the value that should not be there.

This would break as soon as we add a new config option though. Then we'd need to have `("s3", None | Some(_), None | Some(_), None | Some(_))`: the first has to be `None | Some(_)` as before, and the next two have to be `None | Some(_)` because either of them could be set. And that is equivalent to `("s3", _, _, _)` :slightly_smiling_face: